### ISSUE TITLE
io.ascii to parse link and read in full version of tables

### ISSUE THREAD

**Issue by bsipocz** [astropy/astropy#14556]
AAS journals often include full data tables as "electronic only" versions. In these cases the HTML or pdf version of the paper only contains a small subset of the full table, therefor parsing the html/pdf with `io.ascii` leads to an incomplete dataset.

The HTML version of the paper however includes a link to an s3 resource, albeit with an expiration date, so manually harvesting that link and adding in an analysis code is not suitable and reproducible enough as the link will eventually expire.

Feature request:
- parse the HTML, find this link for the full version of the table, read that one in into an astropy Table.

Example for one such Table in a paper is Table 2 in https://iopscience.iop.org/article/10.3847/1538-4357/ac743f#apjac743ft2

cc @hamogu and cc @augustfly as this is as much as an AAS journals issue as an astropy one

### EDIT from @taldcroft 2023-Dec-14

After considerable discussion and some quick prototyping, it appears that the original feature request is not technically feasible due limitations in the upstream data provider(s). In particular, programatic HTTP requests for data tables are seen to be blocked by a Captcha prompt. This is circa Dec 2023 and may change in the future. People encountering this issue are encouraged to read the full comments for the details and a wide-ranging discussion about the future of data in astronomy.

**Comment by pllim** [comment]
Does this issue supersede the following?

* #14346

**Comment by bsipocz** [comment]
No. I mentioned this issue in the comments for that one, but that one is clearly for a bug of parsing a table incorrectly, even though it's fully included in the HTML file, this one is about dealing with a UX "feature" of the HTML version of the AAS journals, where one needs to do one more click to get to the actual table data.

**Comment by augustfly** [comment]
I am super happy to track the issue here, thanks! Let me know if there's something needed from me.

**Comment by bsipocz** [comment]
> Does this issue supersede the following?

The common theme between the two is that there are no tools available that properly parse data tables out of astronomical journal files, and ideally astropy.io.ascii could/should do it out of the box (luckily the field doesn't really have that many journals, so their pecualities can be taken into account even in a generic, though astronomy flavoured, library).

**Comment by pllim** [comment]
Does AAS have a dev to help implement this here? If so, a PR is very welcome!

**Comment by augustfly** [comment]
Sorry, this thread is currently super under-specified, and my reply is going to come across brusque as a result.  I do not know what problem is trying to be solved though I'd love to help enhance reproducibility. Is the core issue defining "reproducible enough" as stated above? Do you have some specs that you can point me to on what that means? 

Here are some questions based on this thread that I hope will help me understand what is desired. 

- Is the problem that there is a short online table and a longer complete one? That's not a problem from our perspective; that has been the case for 23 years. It will not change. While our HTML has changed considerably over time (and all flavors are out there floating around), we are not going to provide HTML versions of useful data when we create MRT versions to replace them. 
- Is it that it takes more than 1 click to get to a table? Again, I don't consider this to be a problem to be solved. I feel that you have some "Journal landing page specifications" that aren't clearly stated, including "how many clicks it should take to get to a data file." A simple summary would help me and probably other journals out a bunch. 
- Is it that the table reference is inconsistently tagged in the HTML in some way so that makes it hard to find the S3 link? In what way? Can it be detailed? I'd love to send an issue up to the publisher to fix their HTML to make these links clearer. Screen-scraping HTML is dangerous business as there's no requirement to freeze the HTML format of any webpage anywhere (You'll find many generations on our site). The underlying problem is not dissimilar from folks who believe that they can use Zenodo URL file links for reproducibility -- nothing about plain URLs is preserved as any requirement of any website. The other issue (14346) about table references linked to this issue isn't from an AAS Journal article.
- Is it that there is an S3 link that you cannot parse? The target of every S3 link in an machine-readable table that parses with format="ascii.cds" in io.ascii 95-99% of the time, throwing only warnings bc astropy.io can't understand log units in our format. The other 1-5% of the time the data is in FITS. Do you want us to help write an S3 link parser to help mime-type identify the content? That might be useful! 
- Is it that HTML renderings of tabular material are inconsistent across all journal websites? That's not a problem to be solved because we have a long-form data type that we create (and lives at the S3 link). It's also not a problem to be solved bc there are no requirements for HTML websites to be stable over time (sorry I keep saying the same thing over and over again!). 
- Also note that all the tables that don't have S3 links have tab-delimited versions that you can uncover in your HTML screen-scraping that we've worked fairly hard on to make cleaner than the HTML (though this is done algorithmically/automatically). For example, here is [Table 1 rendered as tab-delimited](https://iopscience.iop.org/0004-637X/933/2/180/suppdata/apjac743ft1_ascii.txt?doi=10.3847/1538-4357/ac743f). The tab-delimited versions aren't in key'd S3 buckets. The fundamental problem is that authors have near infinite ability to restructure their tables so predicting the header rows and columns is almost impossible, which is why we make the  MRT versions that live on the S3 links.
- Is it that the S3 link expires? Yeah, I am right there with you and really hate this. I've reported it to the publisher as a big a problem for reproducibility, but it is not going to be changing any time soon bc it is some security device put in place by our publisher. Since you are OK with screen-scraping HTML, could you add a call back to get the newest AWSAccessKeyID from the HTML whenever the dataset needs to be re-retrieved? Seems like this must be a common function for key-signed S3 buckets (but I believe you both know 1000% more than I about AWS).

**Comment by augustfly** [comment]
let me add that we do have a procedure to spin up small software projects to build features like this. I just need a lot more information that I feel is currently missing! thanks

**Comment by bsipocz** [comment]
> Is the problem that there is a short online table and a longer complete one? That's not a problem from our perspective; that has been the case for 23 years. It will not change. While our HTML has changed considerably over time (and all flavors are out there floating around), we are not going to provide HTML versions of useful data when we create MRT versions to replace them.
...

no, it's not a problem that there are multiple tables, the problem is that there isn't a machine-ready long table version that doesn't require interactive involvement with the website. astropy.io.ascii is capable of reading in tables from HTML pages, and it indeed finds the short version. The feature request is for making it able to find the s3 link and read it in without user interaction.  (See the referenced notebook above, we did copy pasted the s3 link to read it in, but the link that used to work a week ago has already expired and now throws a 403 making the data acquisition part of the notebook unreproducible, unless one goes back to the HTML page, clicks on the link again and copy-pastes). 
So, yes, I think scraping the s3 uri is the desired solution here, but I would like astropy to do it out of the box, as it affects one of the 3 major astro journals, rather than I as a user need to code it up for my notebook, so does the next user. 

(ultimately, of course, getting these datasets into e.g. vizier would be a nicer solution, or any other reasonably standardized data-sharing platform, that provides easy-to-use, generalized API access. But until then, as a workaround, if astropy provides HTML table reading capabilities, it would be nice if the reader would be fine-tuned to be able to handle astro journals)

**Comment by hamogu** [comment]
I agree with @bsipocz : AAS journals should fix they way tables are stored. Personally, I would prefer to have all suitable AAS journal tables go to VizieE automatically, in the same way that A&A table do. That's not just easier to get through an API (though I admit that reading the "machine-readable tables" is not easy with a machine), it also allows searching through Vizier' I often find A&A articles, because I search some object in VizieR and then see that it's listed not only in GAIA, 2MASS, etc, but also in some table of a much smaller paper that might only deal with 10 sources.

That said, the second best thing is AAS journals to have persistent links for data tables. 

And only last, as a band-aid, should astropy implement a work-around. The goal of `io.ascii` readers it to be able to say "read that article and get me table 2". As Brigitta says, that works in the HTML (thanks for formatting correctly as HTML), but then you only get the top 5 lines. A solution could either be "follow the link to the MRT and parse that" or "parse the full table in HTML". 

Of course, one could also manually download and save the table you get from s3  (That's what I typically do. I then add it to the git repro for the project I'm working on and commit with "downloaded manually from ApJ article XYZ on the date of this commit". Not ideal, but better than nothing.) So, I think it would be a useful, but not high-priority feature for astropy to be able to point it at and ApJ HTML and say "get table 2, follow link to s3, then parse that table".

**Comment by taldcroft** [comment]
I agree completely with @hamogu and @bsipocz above on all points and the priorities. I confess to not being involved in active research and I find it quite surprising that AAS journals are not making every data table available for download with a permalink or (better) via documented API. VizieR seems like an excellent idea for the reason @hamogu mentioned. 

Requiring scientists (or astropy code) to scrape HTML to get at data is a poor solution. That said, some kind of utility to scrape the AAS data download links from HTML could be a near-term workaround. This wouldn't necessarily be part of `io.ascii` but could be a specialized function in `astropy.utils.data` to get links that could then be used in `io.ascii`.

It is also worth noting in the example article (https://iopscience.iop.org/article/10.3847/1538-4357/ac743f#apjac743ft2) that the downloaded "ASCII" first table is not in a machine readable format at all. AAS certainly has the power to enforce that all submitted tables are actually machine readable. There is really no general way for astropy to read this table even if the download link is scraped from the HTML.
```
Table 1 
CLQ Candidate Selection


TDSS Target Type	N _QSOs	N _Cand	% Cand	N _CLQ	% CLQ	
A	17,511	0	0.0	0	0	
B	44,832	1	0.0	0	0	
FES_DE	653	9	1.4	6	0.9	
     ... <snip>
RQS3v	1579	4	0.3	0	0	
Total	64,039	61	0.1	19	0.03	
Note. Target selection methods for our TDSS CLQs. <snip>
```

**Comment by gmzsebastian** [comment]
Thanks to @pllim to pointing me to this issue. I would just like to add that it would be good to also think about how to handle `notes` when importing tables from the standard ApJ Table format. The current way of reading in tables seems to ignore this block of text, which is reasonable given how flexible the contents can be, but I think it would be useful to at least read it as a string that can then be parsed by the user.

**Comment by bsipocz** [comment]
ping @hamogu - this should really be something we get to the bottom of, now that there are paid people. Or just ping AAS to have a fix from them. Either way our science users run into the problem each time a new project is started up making the process more inconvenient than it should be.

**Comment by taldcroft** [comment]
@bsipocz - is there a predictable way to find the table links in the HTML? Maybe it's buried in the comments above but it wasn't obvious if all AAS HTML articles follow any kind of convention. Is there a link regex that can reliably have a MRT on the other end?

**Comment by hamogu** [comment]
Is there any chance that ApJ will ever host the datatables at a persistent URLs? E.g. for the example in this thread, today I get https://content.cld.iop.org/journals/0004-637X/933/2/180/revision2/apjac743ft2_mrt.txt?Expires=1703097468&Signature=omUFGLoQBR1fj-d09swd7LYqfp3VQOv~OtPd45seSnr7SKXUARM2Xl8nXGLAtii8gDeZLSx5Q3Jx1nwhVwQpflD93D7g2vIiUkWaDZL2y0b2IfmkTQEx9DelchOeIb6eoxuKoDyqiBwMySk9Gqlt5F7dnhV0F29bthF7qPc7N8RBtysLdo4yNisKhpuQZ5mX-wt8-gjDhu01k57Jcn92Y2hi5WZ~ilqDykrnxWasxAaDQpocFeVa~WmFocEX9yYM1EF4Xnbs9c6bG6UBZnoP9zoPdYLR7tW4rKH2MYMn5sUpV6wCc1VgKNG2pt9xiAFIPKF6-xIzA8~c1E-dsjKgXQ__&Key-Pair-Id=KL1D8TIY3N7T8
but I know that will expire. It would be much better if there was some canonical way to get the data table,e.g.
`https://iopscience.iop.org/article/10.3847/1538-4357/ac743f/mrt_tab2`. That canonical link could be redirect or something that get to the S3 instance where the table actually sits. If not, then I guess we have to bite the bullet and parse the the ApJ html to find the link of the day.

That's certainly possible, but not nice. (And I will have to think about if that should live in `io.ascii` or maybe in `astroquery.AAS_journals`. I'm afraid that ApJ will break the scheme and then a package that's updated more frequently like astroquery would deal with that much better than astropy itself with it's much longer release cycle./)

@augustfly So, at this stage it would be just good to know if persistent links are never going to happen, might happen in the near future, or if there are other near-term plans  that impact this (e.g. serve ApJ table through Vizier).

**Comment by taldcroft** [comment]
In the original article we see something like this, where the two links seem to point to the same place. 

<img width="661" alt="image" src="https://github.com/astropy/astropy/assets/348089/004621a3-05c6-400b-84a3-9ee6b57a8578">

If we can reliably infer that links like `<a class="btn btn-primary wd-jnl-art-btn-table-data" href="<URL>">Data</a>` are Machine Readable Tables then we can do something even if it is hacky.

So this is for @augustfly - is there a spec for embedding links to tables in the HTML of articles?

**Comment by augustfly** [comment]
Hey, I didn't reply because we have been trying to get canonical links restored in the Journal articles. We met with the publisher in September and stressed the need to fix these data links. We noted that the tab-delimited versions are not blocked with S3 keys so why are the actual data files. 

We are looking at different ways to achieve this, including setting up separate dataset DOIs to store such assets. We completely agree that something looking like persistent links to data are valuable to you (i.e., astropy) but especially to users of the data.

**Comment by augustfly** [comment]
@taldcroft Consider the following question: "Where are the data links that a machine can discover, reliably, for a Zenodo DOI?"  

Right? _They doesn't exist_! You can craft a parser today that correctly guesses a Zenodo data link by screen-scraping and then poking around the resolved URLs to come up with a working algorithm. But no one, especially not the data provider, is going accept a requirement that those data links should persist. It's not a reasonable request because a website's DOM is not a persistent structure. The only thing that is persistent is the DOI to the landing page...

I've been told the "right answer" is something like schema.org tags embedded in the landing page leading to specific assets. This makes sense, but then you'd need a schema implementation that all data providers can follow to make the data links discoverable (regardless of whatever happens to the website's DOM).

**Comment by bsipocz** [comment]
Thanks @augustfly, good to hear that it's on the table.

However, if astropy was able to follow the links the issue of expiring S3 links could be worked around. I suppose the task would be to parse out the link to the full table, as currently, one gets the truncated version.

A similar topic (from the user perspective) is to fix mnras, too, as for those we run into this other problem: https://github.com/astropy/astropy/issues/14346

**Comment by augustfly** [comment]
@hamogu We are also exploring directly making the data av to Vizier rather than only the tables Vizier can ingest. Something closer to A&A (and certainly something faster than we have with our current best-effort arrangement). Those discussions were also started in September and continued at ADASS! (I had many September trips that I could have summarized for this thread since they all have to do with this thread).

**Comment by augustfly** [comment]
@bsipocz I completely understand why there's a strong desire to walk the various journal article websites to pull out results, or to ask that journal websites have some standardization. I'm just not convinced it is wise or possible. It is certainly not something we control at this granular a level (as evidenced by the introduction of the S3 links w/o telling us or considering our use cases). This is a major reason why I'm pushing to move data to something more like a real repository -- for discovery and reuse. Also versioning. And Licensing. And data-specific metadata.

**Comment by bsipocz** [comment]
> This is a major reason why I'm pushing to move data to something more like a real repository -- for discovery and reuse. Also versioning. And Licensing. And data-specific metadata.

And that is all good and nice, but in practice, we need a solution on a much shorter timescale as currently we run into bugs (e.g. a user gets the truncated dataset without realizing it's not the full version, or workflows that used to work suddenly stop working a manually dug up S3 address expires). Pushing everything to vizier/zenodo/whatever can be a nice end goal, but as far as I see it won't happen on the bugfix timescale and therefore a lot of workaround efforts and frustration is generated instead.

**Comment by hamogu** [comment]
I'm not sure I want to get this into astropy on the "bugfix" timescale because then we have to bugfix it every time that some publisher changes their DOM. I forsee test cases that break every few weeks and astropy versions that users run that are broken all the time because they did not install the latest bugfixes. That's why I wonder if "getting the links out of the paper" would be better suited for astroquery, which is set up to release much more frequently. As it stands, I wonder if our recommendation should be "download manually and keep a local copy, that you read your local copy in your workflow".
It's more cumbersome, but published data tables don't change. You only have to download it once by hand - and you have to read the paper by hand at least once anyway to see that that's the table you want, so that's not really any extra work for the user.

Driving this thought to the end, maybe we should REMOVE the ability to read remote tables with `io.ascii` because it's actively harmful at least in this case.

If all ApJ tables went into Vizier, then this would be a non-issue: Astroquery can interface with that already and that works great in all cases I used it so far.

**Comment by bsipocz** [comment]
> You only have to download it once by hand - and you have to read the paper by hand 

And that is not a reproducible workflow to share a project with others. This all came up while we were developing use cases for a science platform while using open science and open data approaches (so nothing is coming from a drawer or a pile of manually curated papers.

As for whether it belongs to astroquery or not, I would think it's not, and it definitely doesn't belong if it doesn't come with a committed maintainer or if any processing happens on the user side. We tried to have ads, but that module is somewhat broken and I'm certain Andy's version is more powerful. So maybe keep papers and paper scanning under that one.

**Comment by hamogu** [comment]
You would put that file into a git repro and curate it that way. Not ideal - but as we see downloading it form ApJ is also not a reproducible workflow in practice at this time.

(I don't know who "Andy" is.)

**Comment by hamogu** [comment]
To, me this is more akin to "download table from Vizier", which currently is under astroquery, so that's where, as a user, I would expect this.

**Comment by bsipocz** [comment]
I assumed this is a well known package: https://github.com/andycasey/ads

And nope, data, especially machine-readable (well, the machine-readable part is what I debate in this issue), etc, doesn't belong in those repositories. Yes, it's a workaround for now, but still, it's not open science if open-access publications cannot be used in a reproducible way.

**Comment by bsipocz** [comment]
And afaik, Vizier is a curated data provider, I don't expect everything will end up in there, but it's CDS who would comment on it properly.

**Comment by taldcroft** [comment]
@augustfly - you answered a question that I did not ask so let me try again. 

If I have the URL to retrieve the HTML of an article, can I parse that recently downloaded HTML and reliably find links to machine readable tables? 

For instance, a notebook might include code to get the third table from this AAS journal URL. I'm assuming that the URL pointing to a journal article is permanent.

**Comment by taldcroft** [comment]
Thinking a bit more, this seems like a job for astroquery. You can think of each journal as a data archive that provides data from journal articles. Every journal or publisher provides such data by a different API that may or may not be documented and that may or may not be stable. Astroquery has the difficult job of trying to retrieve that data as best it can.

The difference here is that the primary search key would be an article and potentially some kind of identifier for which piece of data is requested.  Astroquery already has the basic infrastructure for queries to different data providers, in particular a modular structure with room for extension and a way to document the interface for each provider.

**Comment by augustfly** [comment]
> I clicked through a handful of "Publisher Articles" from ApJ via ADS. A minority of tables are in MRT format. However, most tables appear to be in an ad hoc ASCII table format like below. It's tab separated but has optional elements that would require a parser to make educated guesses. Is anyone aware of guidelines for such tables?
> 

I would be happy to ask IOP for the rules used to convert automatically the XML tables to the tab format. In what format would those rules/transform be most useful to you? 

A review of the typical publication process: 
- Author make LaTeX tables with all kinds of wild & cutesy styling;
- Publishers (mostly by machine) convert the LaTeX to XML tables with numerous internal XML tags to preserve some/most author formatting while expunging other things (like vertical rules);
- XML is transformed to HTML using stylesheets to display the content on the webpage. This transform often embeds HTML-specific style elements in the tables, like line break etc, required to make it appear nicely on the HTML page. 

The AAS (solely among astronomy publishers):
- automatically converts the source XML to a tab-delimited format to remove many/most of these HTML style elements and make the tables easier to parse than pure HTML. We've iterated on this auto-tab process a couple times over the years. Apparently PASP has inherited this functionality on the same platform (see below);
- manually create about 1100-1200 fully machine-readable tables (per annum) from the -- wild guess -- 15-25,000 tables authors publish every year in our journals. Of course the vast majority of the 15-25k tables do not contain actual data; they are often used for style and other display purposes. Manually converting them all or requiring (E)CSV files for all of them would really not be appropriate or efficient I feel. 

We are def exploring additional ways of standardizing author data, but to reiterate my earlier points none of them is related to the content of the HTML article landing page or its DOM.

> It is unfortunate that ApJ / IOP are still using an ad hoc table format like this, but parsing this format would certainly be in scope for an extension to astropy `io.ascii`.
> 

To recap table formats in astronomy journals see the table below. And to be 100% clear, I would welcome a formal analysis that said something like, "We reviewed the HTML tables in the top ten Astronomy Journals and recommend systematic reduction in the styling elements in Journal Table formats to encourage more regular and reusable table data".  Otherwise it is very hard to bend the "deluxetable" cruft + cutesy behaviors we see. I just don't have time to roll that boulder back up the hill. 

| Journal | HTML | MRT | Alt/Tab |
|--------|--------|--------|--------|
| AAS (IOP)| Yes | Yes | Tab |
| A&A | Yes | on Vizier | No |
| MNRAS | Yes | No | No |
| PASP (IOP) | Yes | No | Tab |
| PASJ | Yes | No | No | 
| JGR/AGU | Yes | No | No |
| JHEP | No | No |  No |
| JCAP | No | No | No | 
| Nature | Ha | Ha | Ha |

**Comment by augustfly** [comment]
> @augustfly - you answered a question that I did not ask so let me try again.
> 

I answered I think a better question! To state my point: for an actual data repository, e.g., Zenodo, can you reliably find a URL to a hosted data file and can that URL to the data file be considered permanent? The answer to both is indefatigably, "no."

> If I have the URL to retrieve the HTML of an article, can I parse that recently downloaded HTML and reliably find links to machine readable tables?
> 

Your ability to walk a website's DOM and find anything depends on that DOM being completely transparent. As far as I know it is for IOPScience, the current website's content model is "stable" as far as we understand it, and the MRTs (and tab-versions) are linked consistently in the HTML. Of course the website's structure could change tomorrow to improve accessibility (by which I mean [a11y](https://www.a11yproject.com)) of the content, new data linking tags can be introduced, or you could rightfully be blocked by a robot blocker, which I also hate but exist for valid website security reasons. 

> For instance, a notebook might include code to get the third table from this AAS journal URL. I'm assuming that the URL pointing to a journal article is permanent.

The notebook author's decision to have a URL to a data file on the AAS Journals is a decision of the creator of the notebook. The assumption that any URL anywhere on the web is permanent is incorrect. The DOI of an article is permanent so long as we maintain it, which we are tasked to do. That DOI leads to a landing page providing access to the content of the article. The format and structure of the content of an article at that DOI are required to be neither persistent nor permanent.

**Comment by augustfly** [comment]
> I assumed this is a well known package: https://github.com/andycasey/ads
> 
> And nope, data, especially machine-readable (well, the machine-readable part is what I debate in this issue), etc, doesn't belong in those repositories. Yes, it's a workaround for now, but still, it's not open science if open-access publications cannot be used in a reproducible way.

My vision for "open and reproducible science" is one where you can ask the article DOI for all the "data" related to an article and get back a series of additional dataset DOI(s) that can be inspected for their contents, including schema-discoverable data items/files.  I think this is the right "open science" path forward. I feel it is the right way because this path is intentional, flexible, and designed for machine-discovery and reuse. 

I feel like a broken record saying again and again that HTML journal articles were designed for display of content not any of the poorly specified requirements on this thread. The right "open science" path forward is not one that make these sorts of HTML tagging demands of the current publishing platforms to provide persistent data links. The logic is undermined by the reality that neither websites nor their URLs nor their DOMs have any requirements for persistence -- anywhere ever full stop! If we want some of those things then we have to build it that way -- see my first paragraph as a path forward (or so I assert!). 

Lastly, if a group of astronomers and developers have analyzed the _current_ system and have a PEP or other document that supports the assertions about what is open and reproducible science in this post then please provide it as a link. I would like very much to read that and perhaps be swayed from my current beliefs! 😃

**Comment by taldcroft** [comment]
@augustfly - I understand your vision, and that is certainly a good thing to pursue. Given the ambitious nature and scope of that vision, the time scale for that to become a reality is not clear.

What this issue is addressing is what we can do *today* to make research a *little more* reproducible and convenient. Perfect is the enemy of good. Poorly specified or non-existent requirements are a core part of astrophysical research (including much of astropy). So the goal is to do the best we can, and from what I can see *today*, HTML is the only option.

With that we accept that any solution might just break and then the software would need to adapt to the new situation.

> As far as I know it is for IOPScience, the current website's content model is "stable" as far as we understand it, and the MRTs (and tab-versions) are linked consistently in the HTML.

I'll confess to being a complete newbie in the document world, so I was just pleasantly surprised to see that `https://doi.org/10.3847/1538-4357/accffb` redirects to the ApJ HTML (where `10.3847/1538-4357/accffb` is the DOI for some random ApJ article). I expect most people on this thread already knew that. That implies, at least for ApJ, a path from DOI to HTML that *might* work most of the time?

**Comment by taldcroft** [comment]
@augustfly - if you are interested in making a concrete incremental improvement toward your "open and reproducible science" vision, you might start with trying to convince IOP to publish **all** data tables in MRT. As I mentioned it appears that a large fraction of article tables are in an ad hoc tab-delimited format which requires guessing in order to parse.

**Comment by augustfly** [comment]
> @augustfly - I understand your vision, and that is certainly a good thing to pursue. Given the ambitious nature and scope of that vision, the time scale for that to become a reality is not clear.
> 
> What this issue is addressing is what we can do _today_ to make research a _little more_ reproducible and convenient. Perfect is the enemy of good. Poorly specified or non-existent requirements are a core part of astrophysical research (including much of astropy). So the goal is to do the best we can, and from what I can see _today_, HTML is the only option.
> 
> With that we accept that any solution might just break and then the software would need to adapt to the new situation.
> 
> > As far as I know it is for IOPScience, the current website's content model is "stable" as far as we understand it, and the MRTs (and tab-versions) are linked consistently in the HTML.
> 
> I'll confess to being a complete newbie in the document world, so I was just pleasantly surprised to see that `https://doi.org/10.3847/1538-4357/accffb` redirects to the ApJ HTML (where `10.3847/1538-4357/accffb` is the DOI for some random ApJ article). I expect most people on this thread already knew that. That implies, at least for ApJ, a path from DOI to HTML that _might_ work most of the time?

The developer answer is to use the DOI API to "ask" for the resolved URL of the DOI (the fact that the DOI is embedded in the IOP URL is a coincidence). I use `from habanero import Crossref` for all my DOI API tasks.

**Comment by augustfly** [comment]
> @augustfly - if you are interested in making a concrete incremental improvement toward your "open and reproducible science" vision, you might start with trying to convince IOP to publish **all** data tables in MRT. As I mentioned it appears that a large fraction of article tables are in an ad hoc tab-delimited format which requires guessing in order to parse.

Unfortunately IOP doesn't have a role here. It is an AAS goal to provide MRTs for valuable data.  We have a team of 3 people who full time turn author's latex / data into the MRT format. We create about 1100-1200 MRTs per year, which we estimate is about 20-25% of the total tables out there. Not all tables have data though I think that is important to point out. See also:

Muench A. The roles of data editors in astronomy. Sci Ed. 2023;46:8-10.
https://doi.org/10.36591/SE-D-4601-04

**Comment by hamogu** [comment]
I agree with most of the above. We need both: The push for a better long-term solution as outlined by @augustfly and a more short-term "better than now, but not really good" solution as @taldcroft wrote. The latter can then hopefully be retired at some point.

DOI -> HTML works for most journals (at least those where don't have to go through a paywall log-in to see the full article). The issue is that the HTML representation might change at any time, as @augustfly pointed out. In some cases it has in the past in other cases, the same journal has actually different html, depending on when an article was published (e.g. A&A changed the design of the HTM a few years ago, but they kept the old articles stable. That means that anyone who relied on the structure of the old HTML still can for the old articles, but it also means the an "A&A parser" has to understand multiple different HTML layouts if you want it to work on all A&A articles).

So, how do we move from here?
1. Based on all that's we said above I will 
  - (a) implement a POC function to download an electronic table from an ApJ DOI using the current HTML of ApJ. That table might be MRT or some other format. Once we have that function, we can decide where it's going to live in the future. I am strongly opposed to put it into `astropy.io.ascii`.
  - (b) Look at the tab-delimited tables and see if I find a way to parse a good fraction of them with `io.ascii`. It's unlikely that there is a way to parse all of them automatically. Not ideal, but a step forward: In @bsipocz "reproducible science" scenario, at least you get the same table every time from (a), so if you need to write custom parsing code for a project, you can and still get the same results.
  - (c) look at MNRAS. I feel Nature is hopeless and A&A is less pressing, because at least the big and important tables from A&A are on Vizier. 
2. We need white paper or something as suggested by @augustfly and ideally that's written by someone not employed by ApJ/AAS. I suggest to implement the short-term solutions first, so that we can point to them in that white paper and say "it's not a good long-term solution". That brings me to a follow-up question for @augustfly: Assuming we improve MRT writing in astropy (I know I talked about that with you in a different context) and authors submit all their tables as MRT, would you be able to publish that MRT instead of the tab-delimited versions?

It is planned that MIT will receive a sub-grant from Astropy's NASA grant to pay for some of my time early next year; that grant is currently stuck somewhere in the administration between NASA, NumFOCUS, and MIT. I don't have to strictly wait for that, but I won't get to this before the end of the year in any case.

**Comment by taldcroft** [comment]
OK, NEVER MIND. I just started working on a quick POC and discovered that the ASCII download is protected by a Captcha. 

AAS journals and/or their data provider are explicitly blocking the ability to download data tables programmatically. So as far as I can tell we are done with incremental progress reproducible science *today*. If that is correct we can probably close this issue.

Argh.

**Comment by taldcroft** [comment]
https://gist.github.com/taldcroft/00911ca21cddf43f82c0c3365ea3b219

**Comment by augustfly** [comment]
> AAS journals and/or their data provider are explicitly blocking the ability to download data tables programmatically. So as far as I can tell we are done with incremental progress reproducible science _today_. If that is correct we can probably close this issue.

The AAS never asked for the captcha (which is intermittent, depends on multiple requests from the same IP, and is for the entire article not simply the data table or other assets); nor did we ever ask for the S3 keyed links. 

We prob want a platform with good uptime and protected from DDOS/mass scraping etc, but these two solutions totally suck -- I believe I see that captcha more than anyone since I do regular audits of the HTML content. It's ridiculous honestly.

**Comment by gilleslandais** [comment]
Because VizieR was cited, I make a short report about the CDS and AAS cooperation.

For a long time now, we have an active workflow to curate ant put in line data from AAS journal.
The CDS makes a selection of the articles and prioritize in it the MRT tables (the same format used with A&A). Other tables in HTML may complete the catalogues when they are considered interesting for VizieR.
Curation needs investment and the MRT prepared by AAS allows us to enrich the database.

With the advent of open sciences, we need to think about new possibilities. For instances, improving Data citation or improving the hosting of authors’ data. The topics are discussed with AAS, in previous meeting (cited by Gus), and it will continue in the next AAS conference in a more widest audience. 
All of these initiatives needs investment and preliminary research – So it's a long-term job!
Note, the open science feedbacks and intiatives are also discussed in the IVOA conferences (you are welcome) in the IVOA DataCP session

**Comment by augustfly** [comment]
> 2\. We need white paper or something as suggested by @augustfly and ideally that's written by someone not employed by ApJ/AAS. 

@hamogu Completely agree on all aspects. If there's not wide buy in and external pressure then it's very hard to move these systems along. Just don't leave the publishing platforms out -- so much of this thread is about ideas of what these platforms are supposed to do when those requirements literally never existed before. It would be like someone writing an angry thread about how unbelievable it is that astropy doesn't support [SPICE kernels](https://naif.jpl.nasa.gov/naif/data.html)! I wouldn't be surprised if your reactions wouldn't be similar to mine right now, which are a mix of I want to support everything you want to do and these demands are out of scope / poorly formed.

I'd also advocate including arXiv as they are preparing HTML versions of preprints and that version of the tabular data may also be useful. 


> I suggest to implement the short-term solutions first, so that we can point to them in that white paper and say "it's not a good long-term solution". That brings me to a follow-up question for @augustfly: Assuming we improve MRT writing in astropy (I know I talked about that with you in a different context) and authors submit all their tables as MRT, would you be able to publish that MRT instead of the tab-delimited versions?

@hamogu Yes. There is no "tab-delimited" version when there is an MRT (unless there's something buried in the HTML I can't see). The "Data" link is replaced by the MRT link. (This is why I found the earlier reported behavior of retrieving incomplete data so confusing. Sure the "HTML" table is a stub, but the data link is always present and always leads to a complete version of the data).  

And I think the astropy writer is really pretty close after I looked at it again. There's just one minor header formatting problem that causes us to freak out, and despite the nice pressure from ppl like Kelle and Erik, I'm still afraid of the PR process here. There's also a couple functional steps in there (log units have units please; not everything should be saved as double precision floats) that have wider astropy implications, but we can ignore those for now.

**Comment by hamogu** [comment]
@augustfly Please don't take this thread as a specification of what we want or an attack of "why does ApJ not do this?". Instead, I would like you to read it as a brain-storming of what we can do now and what we might want in the long run. The original question as posed by @bsipocz is essentially "How can we work best with what' already on the web?"
In that, there is no request for you to do anything, we just want to work with the existing resources. From there, I've just been trying to understand what your thoughts  future plans are - as you point out, parsing the HTML is fragile, so I just wanted to know if e.g. you plan a change in the near future anyway. That's why the request is not "specific" - it's "what can we do with what's there already".

But then, as you say, it would be good to develop a specific request - and clearly that's a longer process that involves users, data providers, publishers, etc.  - and in that part astropy might play a role again in giving users the opportunity to provide e.g. well-formed MRT tables to the journals, which would make it easier for the journals to serve those again for download. 

One takeaway I have from this discussion is that we, the astropy project, should probably be more engaged at IOVA, ADASS and similar forums. In the past, we've just been consumers - we try to provide readers for whatever format the authorities come up with, but maybe it would make our live easier if some of us took a more active role in that process, too.

**Comment by bsipocz** [comment]
@taldcroft - before considering closing this issue as non-implementable*, maybe adding a warning for truncated tables is still in scope and totally doable technically?
 

* even if that's the case, I would argue that keeping it open still serves the purpose that this was considered and instead of not-implementable I would consider it not-yet-implementable, after all there are gazillion issues like this still open

**Comment by hamogu** [comment]
It might be implementable - you don't (usually) hit the captcha if you only request one table per day. It just means that development is a little tedious because we can't try it out several times in a row and we can't run it on CI. So, I think we'll leave it open.

**Comment by augustfly** [comment]
@hamogu Really grateful for that contextualization! I have to step off here for now bc I have a big internal "data" meeting tomorrow in fact about some of these topics, and I'm behind! But I'll say this about change: funders are pressing authors to put data into DOI-based repositories in the name of [F#*!](https://www.go-fair.org)ness. We see this happening (exponentially) whether we (the AAS) do nothing or embrace it. 

My personal preference is to embrace the move to data repositories, and I think this would include moving data assets off the publisher landing page and to a dedicated data partner / repository -- optimally one that can provide the most powerful, reusable, and discoverable formats and APIs for that data. Again my personal goal/preference.

**Comment by augustfly** [comment]
^ ofc that vision does nothing for the data currently in the Journals! 😆 😮

**Comment by taldcroft** [comment]
@bsipocz - agreed to leave this open. I added some commentary in your original description to help anyone that might stumble on this issue and maybe inspire action. I also added `wont-fix` and `Upstream Fix Requires` labels. I think those apply for the narrowly defined feature request.

**Comment by taldcroft** [comment]
@hamogu - I hit the captcha on my first try and every try afterward. Even if it occasionally works, I think that experience would preclude this approach being used in an Astropy Project core or coordinated package.

**Comment by bsipocz** [comment]
@taldcroft - What about a warning, can it be parsed out whether a table is a full version or just a truncated one?

**Comment by taldcroft** [comment]
>  What about a warning, can it be parsed out whether a table is a full version or just a truncated one?

I think that any code of that nature will be fragile and is getting very niche. I was enthusiastic enough about reliably downloading all the tables (as "proper" ASCII tables), but scraping the HTML which may or may not be the full table is not inviting (to me).

**Comment by tactipus** [comment]
hey everyone! can we not just find a captcha tool that can help us get the data we need? I've worked with making crawlers before. I could look into it. granted, it's not a sustainable long term solution. Might I suggest that we simply save the page as a PDF to be saved somewhere for later parsing? Just an idea.

**Comment by hamogu** [comment]
If we want just one table, we can go and download it by hand and save it somewhere - in fact that's what I do when I need a table for one of my own projects. I go to ApJ by hand, copy (or download, if MRT) into a local file and work from there.

The goal of this issue is to make EVERY table in EVERY article accessible through automated means, and for that we would have to have a mechanism to beat the captcha every time or we would have to download every table from every article and store a copy in some other data repository. I don't think either of those is a good solution and would probably (just a guess, because why else would they put in a captcha?) violate the terms of service of the ApJ website.

**Comment by tactipus** [comment]
the only solution, then, would be to reach out to ApJ & try to work something out. This thread doesn't seem to indicate that we've done that. Otherwise, any solution would violate their ToS, most likely.

**Comment by hamogu** [comment]
We did reach out to ApJ. @augustfly is the data editor at ApJ, see his comments in the thread above.

**Comment by tactipus** [comment]
thanks! sorry, I am not used to github threads. I put in "captcha" in the search bar for this page to make sense of it. OK...so, there seems to be a contention forming around whether this project's specifications are clear enough, is that right? Among other things :x

**Comment by hamogu** [comment]
I agree with most of the above. We need both: The push for a better long-term solution as outlined by @augustfly and a more short-term "better than now, but not really good" solution as @taldcroft wrote. The latter can then hopefully be retired at some point.

DOI -> HTML works for most journals (at least those where don't have to go through a paywall log-in to see the full article). The issue is that the HTML representation might change at any time, as @augustfly pointed out. In some cases it has in the past in other cases, the same journal has actually different html, depending on when an article was published (e.g. A&A changed the design of the HTML a few years ago, but they kept the old articles stable. That means that anyone who relied on the structure of the old HTML still can for the old articles, but it also means the an "A&A parser" has to understand multiple different HTML layouts if you want it to work on all A&A articles).

So, how do we move from here?
1. Based on all that's we said above I will
    - Look at the tab-delimited tables and see if I find a way to parse a good fraction of them with `io.ascii`. It's unlikely that there is a way to parse all of them automatically. Not ideal, but a step forward: In @bsipocz "reproducible science" scenario, at least you get the same table every time, so if you need to write custom parsing code for a project, you can and still get the same results.
    - look at MNRAS. I feel Nature is hopeless and A&A is less pressing, because at least the big and important tables from A&A are on Vizier.
2. We need a white paper or something as suggested by @augustfly and ideally that's written by someone not employed by ApJ/AAS. I suggest to implement the short-term solutions first, so that we can point to them in that white paper and say "it's not a good long-term solution". That brings me to a follow-up question for @augustfly: Assuming we improve MRT writing in astropy (I know I talked about that with you in a different context) and authors submit all their tables as MRT, would you be able to publish that MRT instead of the tab-delimited versions?

**Comment by augustfly** [comment]
> That brings me to a follow-up question for @augustfly: Assuming we improve MRT writing in astropy (I know I talked about that with you in a different context) and authors submit all their tables as MRT, would you be able to publish that MRT instead of the tab-delimited versions?


@hamogu IOP always replaces the auto-created, tab-delimited versions with MRTs if we can supply them. Sometimes when I work on an article that has some long "stubbed" MRTs, I also create MRT equivalents of all the other short, unstubbed, tables that have data. For example, yesterday, I finished one where there were 4 "stubbed" MRTs, and I went ahead and created 16 additional MRTs for the other 16 unstubbed tables that had object-oriented data. There were a couple tables I did not create MRTs for bc they did not contain data (AFAIAC).  We started making these extra "unstubbed" MRTs (I call them "joins" bc they often contain a name or key that makes them joinable to the other tables) at Vizier's request about 7 years ago. 

So "yes".

Also, we have been using this thread to try to obtain changes from our publisher along the lines of this effort. From the last round of posts we received a promising "action" from IOP to remove the S3 links from any data links we specify -- prob after the AAS meeting. They want more information on the captcha problem. @taldcroft if you are willing given the frustration this has caused, can you summarize in your words the captcha experience as listed in your gist? 

I would have said that you want to take an article DOI, walk the HTML to find all data links, provide a summary of those links to the user to "pick" the one to download, use, and parse or summarize the contents of each data link.  If so then I think the captcha problem comes in the multiple repeat "hits" to the server -- the initial HTML parse went OK, but the captcha was hit when an individual data file was downloaded.  This language  -- "multiple or sequential repeat hits to the server triggering the captcha" is IOP's language. Any clarification you can provide may help us argue for removing that problem. I'm hopeful at least.

**Comment by ManonMarchand** [comment]
Might also add that some AAS tables are already in VizieR. Not all of them but you might want to check? 
![image](https://github.com/astropy/astropy/assets/16650466/09ee29a7-c73e-491d-b35b-18b9de77ce63)

**Comment by hamogu** [comment]
Most tables are actually NOT on VizieR - and those are the ones we are talking about in this thread, for example look at this one: https://iopscience.iop.org/1538-3881/166/4/148/suppdata/ajacefbet4_ascii.txt?doi=10.3847/1538-3881/acefbe

The ones are ARE on VizieR are the ones that are already solved: They can be downloaded with astroquery and read with the CDS reader in `astropy.io.ascii`.

**Comment by hamogu** [comment]
But to be clear: It is great that ApJ (and other journals) go through the effort of putting so much of their data into an archive with an API that can be queried and that serves the data in a format that's reasonably well documented and has been stable for decades. 
Thanks to the journal data editors for that effort!

**Comment by hamogu** [comment]
And thanks to CDS for hosting and providing that service (and, this goes well beyond the scope of this thread) for their help on astroquery and the APIs to make that accessible.

**Comment by ManonMarchand** [comment]
There is a tiny bit more than the ones that load directly. Some authors do the effort to submit their data directly to CDS to see if the data could be accepted into VizieR, and some additional tables are selected by the team to be standardized. I was just suggesting that any parser for the html could make use of that and test if this is already existing in MRT somewhere.

**Comment by taldcroft** [comment]
> I would have said that you want to take an article DOI, walk the HTML to find all data links, provide a summary of those links to the user to "pick" the one to download, use, and parse or summarize the contents of each data link. If so then I think the captcha problem comes in the multiple repeat "hits" to the server -- the initial HTML parse went OK, but the captcha was hit when an individual data file was downloaded. This language -- "multiple or sequential repeat hits to the server triggering the captcha" is IOP's language. Any clarification you can provide may help us argue for removing that problem. I'm hopeful at least.

@augustfly - that's basically it. I hit a captcha on my very first try to download a data table in that notebook. I just tried to re-run it and got a captcha just trying to download the original article HTML. So I would say that any naive programmatic access to IOP content at this time is **unreliable**.  By naive I mean that expert spammers my understand how to fool the IOP servers but I do not -- my simple `requests.get(url)` often results in a captcha.

**Comment by hamogu** [comment]
> I was just suggesting that any parser for the html could make use of that and test if this is already existing in MRT somewhere.

That's a great idea; I'm just not sure how to do that in most use cases. It depends on what people put into the parser, e.g. if they download that html file and work on that locally, it's going to be hard to scan the HTML for identifying marks, go out to VizieR, look for something similar, ans ten work on that. And by "hard" I mean, it requires a lot of hand-coding to make it work for the major astronomical journals where the html all looks different etc - that's probably beyond the scope of `astropy.io.ascii` and better done at e.g. ADS. 

However, the low-hanging fruit is probably to point that out in the docs, something like "many tables in this format can also be found in that format which is much easier to parse - looks at the ADS and see if there is a link to format X" or something like that.

**Comment by augustfly** [comment]
@taldcroft thanks; really appreciate it. I had never gotten the captcha clicking around on data links inside an article before, but I very often get it picking an article to read (bc I audit/review 5-10 articles at a time, which I guess IOP thinks is abnormal).

**Comment by pllim** [comment]
tl;dr -- I see a won't fix label. Does that mean we should close this issue as won't fix?

**Comment by hamogu** [comment]
While we won't fix the full issue at this time, there are still aspects of this discussion that we can work on (for example, I'm trying to see if I can make an ASCII reader that can read a large majority of the ApJ "ASCII" format tables.

Also, the AAS journals (see comments above) might change the way those tables are referenced, so some of the issues that stop us from fixing this (like the captchas or very short-lived S3 links) might go away in the forseeable future. If that happens, we'll remove the "won't fix" label. Thus, I think this is one o those (slightly annoying) cases that have a "won't fix" but should stay open.

**Comment by augustfly** [comment]
Hi. 

Around the time of my last comment we (the AAS) issued a request to IOP to remove the keyed S3 links and fix the CAPTCHA problems for data files. 

IOP removed the keyed S3 links to all data files around January 12th of this year, and we met with IOPScience last week to get some additional details on how they implemented static links to the data files. I'm working on trying to summarize how the links now work, which I will post soon. Here are four articles I'm using as examples. 

Article 1: https://doi.org/10.3847/1538-4365/aceda8 (Stubbed MRT and data behind the figure. Also versioning)
Article 2: https://doi.org/10.3847/1538-3881/aced3c (Tar file of other stuff)
Article 3: https://doi.org/10.3847/1538-4357/acb432 (Stubbed and unstubbed MRTs; also useful for the breadth of “tables” authors make)
Article 4:  https://doi.org/10.1086/340554  (pre-IOP data links; yes, I'm using my article)


The CAPTCHA problem hit when crawling an article is not fixed; they are wondering if there are URL request hints that could be negotiated to sort of "white list" astropy access. We imagined that the CAPTCHA problem would manifest differently when developing a tool, running test/CI, and in a user context. We also wondered if with a stable link structure, crawling an article is still necessary. 

What would a conversation between developers here and IOPScience look like -- to work around the captcha issue? I’m prob not doing the best job ferrying developer-level conversations across the Styx. I doubt they are going to come on GitHub to talk this through on an issue thread, but we could ask.

**Comment by hamogu** [comment]
This is fantastic news and thank you for pushing on that. I'm happy to engage with IOP if you think that helps, but what form is the most efficient to do that? If they are willing to look here, that would allow all interested parties to be on the same page, but if you believe that, e.g. an email conversation or a zoom call with 2-3 users/astropy developers is more likely to convince them, then we can organize that, too (and ask here for volunteers to join that call or to provide any additional input).

**Comment by augustfly** [comment]
@hamogu Cool! It is unclear to me how best to move forward on the CAPTCHA part. For instance, I just received the  following request from IOP that has an efficiency I can't adjudicate: 

> Would you be able to get some more information for me, possibly the easiest next step would be for the people who are running the automated requests  to do it again and let me know:
> 
> - Time of the requests
> - URL being accessed (if possible)
> - IP address that the software is being run from
> 
> We can then inspect our log files and hopefully can find some kind of signature in them that we can use to disable the CAPTCHA for those requests

One way I could try to do this is to take @taldcroft's notebook <https://gist.github.com/taldcroft/00911ca21cddf43f82c0c3365ea3b219> and run it locally until I hit a captcha and then send my information to IOP. 

This presumes that the notebook will look anything like production code or even CI/test requests. But it is a start I'm happy to make with IOP to keep them moving forward.

**Comment by hamogu** [comment]
@augustfly I would suggest to do exactly what you suggest: Run the notebook and see if that helps. My best guess is that the only distinguishing feature will be the `user-agent` in the header. For requests using e.g. the `urllib` from Python, it's going to be set to something like `"python-requests/2.28.2"`; if you use a real browser, it's going to be set to something like `'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.95 Safari/537.36'`. If IOP is willing to open it to all Python requests, then it does not matter how exactly the script is written, because all requests using the standard libraries will have the same user-agent (in that case, one might wonder what the Julia users will say, but Python is by far the largest community that has asked for this, so that would still be a good place to start.). But maybe there is something else in their log files that might turn out to be useful. I'm not a web-developer...

**Comment by augustfly** [comment]
Super. I ran the test, triggered the captcha and sent the necessary information to IOP. I was able to get 1 data file before being blocked. Then then nothing.

**Comment by taldcroft** [comment]
Wow, great to see some real progress here, and I'm glad that my quick notebook ended up being useful!

About allowing a particular `user-agent`, I'm a little skeptical about this solution. This would amount to security through obscurity, guessing that folks generating unwanted downloads either don't use Python already or that they can't figure out that Python scraping is allowed. 

Maybe more to the point is understanding the underlying reason for having a CAPTCHA. One presumes they don't want to allow people to just crawl and download the entire data archive? Or maybe this is protection from DOS attacks?

**Comment by hamogu** [comment]
> About allowing a particular user-agent, I'm a little skeptical about this solution.

It's up to IOP, not us to decide what they want to do. I was just speculating about what differences they may see in their logs and point out that at least in some characteristics (like the user-agent) the notebook will indeed look like production code or even CI/test requests, so it's not a futile exercise to answer IOP's request that @augustfly listed at https://github.com/astropy/astropy/issues/14556#issuecomment-1915486513 that at this stage.

**Comment by augustfly** [comment]
We are experimenting with `referer` passthrough. It's sort of working. I'll post my revision of @taldcroft's notebook as soon as IOP has done some more thinking if this passthrough is legit or not. 

And yes, all forms of DDOS and other platform service slowdowns/interruptions are the reasons for the CAPTCHAs. That is what we are told anyway.
